rclcpp/include/rclcpp/memory_strategy.hpp

```diff
@@ -38,137 +36,171 @@ namespace memory_strategy
  */
 class MemoryStrategy
 {
+public:
+  RCLCPP_SMART_PTR_DEFINITIONS_NOT_COPYABLE(MemoryStrategy);
+  using WeakNodeVector = std::vector<std::weak_ptr<node::Node>>;
 
-  friend class executor::Executor;
+  // return the new number of subscribers
+  virtual size_t fill_subscriber_handles(void ** & ptr) = 0;
 
-public:
-  RCLCPP_SMART_PTR_DEFINITIONS(MemoryStrategy);
-
-  /// Borrow memory for storing data for subscriptions, services, clients, or guard conditions.
-  /**
-   * The default implementation stores std::vectors for each handle type and resizes the vectors
-   * as necessary based on the requested number of handles.
-   * \param[in] The type of entity that this function is requesting for.
-   * \param[in] The number of handles to borrow.
-   * \return Pointer to the allocated handles.
-   */
-  virtual void ** borrow_handles(HandleType type, size_t number_of_handles)
+  // return the new number of services
+  virtual size_t fill_service_handles(void ** & ptr) = 0;
+
+  // return the new number of clients
+  virtual size_t fill_client_handles(void ** & ptr) = 0;
+
+  virtual void clear_active_entities() = 0;
+
+  virtual void clear_handles() = 0;
+  virtual bool collect_entities(const WeakNodeVector & weak_nodes) = 0;
+
+  /// Provide a newly initialized AnyExecutable object.
+  // \return Shared pointer to the fresh executable.
```

Comment:
If you want doxygen to pick this up I think you need the `\return` to be within a `/* */`. But I'm not 100% sure.